sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java

```diff
@@ -0,0 +1,169 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.spanner;
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.spanner.Type;
+import com.google.common.base.Preconditions;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Encapsulates Cloud Spanner Schema.
+ */
+class SpannerSchema implements Serializable {
+  private final List<String> tables;
+  private final Map<String, Integer> tableIndex;
+  private final Map<String, List<Column>> columns;
+  private final Map<String, List<KeyPart>> keyParts;
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Builder for {@link SpannerSchema}.
+   */
+  static class Builder {
+    private final Map<String, List<Column>> columns = new HashMap<>();
+    private final Map<String, List<KeyPart>> keyParts = new HashMap<>();
+
+    public Builder addColumn(String table, String name, String type) {
+      addColumn(table, Column.create(name.toLowerCase(), type));
+      return this;
+    }
+
+    private Builder addColumn(String table, Column column) {
+      List<Column> list = columns.get(table);
+      if (list == null) {
+        list = new ArrayList<>();
+        columns.put(table.toLowerCase(), list);
+      }
+      list.add(column);
+      return this;
+    }
+
+    public Builder addKeyPart(String table, String column, boolean desc) {
+      List<KeyPart> list = keyParts.get(table);
+      if (list == null) {
+        list = new ArrayList<>();
+        keyParts.put(table.toLowerCase(), list);
+      }
+      list.add(KeyPart.create(column.toLowerCase(), desc));
+      return this;
+    }
+
+    public SpannerSchema build() {
+      return new SpannerSchema(columns, keyParts);
+    }
+  }
+
+  private SpannerSchema(Map<String, List<Column>> columns, Map<String, List<KeyPart>> keyParts) {
+    this.columns = columns;
+    this.keyParts = keyParts;
+    tables = new ArrayList<>(columns.keySet());
+    tableIndex = new HashMap<>(tables.size());
+    Collections.sort(tables);
+    for (int i = 0; i < tables.size(); i++) {
+      tableIndex.put(tables.get(i), i);
+    }
+  }
+
+  public int getTableIndex(String tableName) {
```

Comment:
Should this really be part of a "schema" abstraction? It feels like more like something that helps some callers do some things with the schema more efficiently, and it seems arbitrary to declare that SpannerSchema will help this particular type of optimization. I'd prefer if callers did this themselves. Especially because the index is stable only within a single snapshot of the schema, so it's kind of a dangerous optimization to include for potentially unsuspecting callers.

Reply:
Done.